From c6eb74c7f3c0dd421c61d7b4d189b496c811c534 Mon Sep 17 00:00:00 2001
From: Yu Liao <liaoyu15@huawei.com>
Date: Thu, 9 Feb 2023 16:46:38 +0800
Subject: [PATCH 884/905] [Huawei] Revert "fs/proc/uptime.c: Fix idle time
 reporting in /proc/uptime"

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188360

--------------------------------

This reverts commit 02989ce4b4366c5d0779978201296188cdc9ded8.

Commit 02989ce4b436 ("fs/proc/uptime.c: Fix idle time reporting in
/proc/uptime") intruduces a issue that idle time in procfs /proc/uptime
goes backwards. Reverts it to address this issue.

 Conflicts:
	include/linux/kernel_stat.h
	fs/proc/uptime.c

Fixes: 02989ce4b436 ("fs/proc/uptime.c: Fix idle time reporting in /proc/uptime")
Signed-off-by: Yu Liao <liaoyu15@huawei.com>
---
 fs/proc/uptime.c | 14 +++++---------
 1 file changed, 5 insertions(+), 9 deletions(-)

diff --git a/fs/proc/uptime.c b/fs/proc/uptime.c
index deb99bc9b7e6..5a1b228964fb 100644
--- a/fs/proc/uptime.c
+++ b/fs/proc/uptime.c
@@ -12,22 +12,18 @@ static int uptime_proc_show(struct seq_file *m, void *v)
 {
 	struct timespec64 uptime;
 	struct timespec64 idle;
-	u64 idle_nsec;
+	u64 nsec;
 	u32 rem;
 	int i;
 
-	idle_nsec = 0;
-	for_each_possible_cpu(i) {
-		struct kernel_cpustat kcs;
-
-		kcpustat_cpu_fetch(&kcs, i);
-		idle_nsec += get_idle_time(&kcs, i);
-	}
+	nsec = 0;
+	for_each_possible_cpu(i)
+		nsec += (__force u64) kcpustat_cpu(i).cpustat[CPUTIME_IDLE];
 
 	ktime_get_boottime_ts64(&uptime);
 	timens_add_boottime(&uptime);
 
-	idle.tv_sec = div_u64_rem(idle_nsec, NSEC_PER_SEC, &rem);
+	idle.tv_sec = div_u64_rem(nsec, NSEC_PER_SEC, &rem);
 	idle.tv_nsec = rem;
 	seq_printf(m, "%lu.%02lu %lu.%02lu\n",
 			(unsigned long) uptime.tv_sec,
-- 
2.17.1

